[backport-2.12] 📦 Integrate manpage builds into PEP 517 build backend

webknjaz opened the pull request:
This patch creates a thin wrapper around the `setuptools`' PEP 517 build backend in-tree. It features an ability to request generating the manpage files in the process of building a source distribution. This toggle is implemented using the `config_settings` mechanism of PEP 517.
One must explicitly pass it a CLI option to the build front-end to trigger said behavior. The packagers are expected to use the following call:

    python -m build --config-setting=--build-manpages

This option has no effect on building wheels.

🧪 The change includes integration tests

This test runs building and re-building sdists and wheels with and without the `--build-manpages` config setting under the oldest-supported and new `setuptools` pinned.

It is intended to preserve the interoperability of the packaging setup across Python runtimes.

An extra smoke test also verifies that non PEP 517 interfaces remain functional.

PR #79606

Co-authored-by: Matt Clay <matt@mystile.com>
(cherry picked from commit 56036013cd09db0f1b20402a69be44345ddafb22)

##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
$sbj
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Maintenance Pull Request
- Packaging Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`packaging/pep517_backend`
`pyproject.toml`
`setup.cfg`

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```console
$ python -m build --config-setting=--build-manpages --sdist
[...]
* Creating virtualenv isolated environment...
* Installing packages in isolated environment... (setuptools >= 39.2.0, wheel)
* Getting dependencies for sdist...
running egg_info
writing lib/ansible_core.egg-info/PKG-INFO
writing dependency_links to lib/ansible_core.egg-info/dependency_links.txt
writing entry points to lib/ansible_core.egg-info/entry_points.txt
writing requirements to lib/ansible_core.egg-info/requires.txt
writing top-level names to lib/ansible_core.egg-info/top_level.txt
reading manifest file 'lib/ansible_core.egg-info/SOURCES.txt'
reading manifest template 'MANIFEST.in'
adding license file 'COPYING'
writing manifest file 'lib/ansible_core.egg-info/SOURCES.txt'
* Installing packages in isolated environment... (docutils, jinja2, pyyaml, straight.plugin)
* Building sdist...
running sdist
running egg_info
[...]
Writing ansible-core-2.15.0.dev0/setup.cfg
Creating tar archive
removing 'ansible-core-2.15.0.dev0' (and everything under it)
Successfully built ansible-core-2.15.0.dev0.tar.gz
```

webknjaz committed:
[backport-2.12] 📦 Integrate manpage builds into PEP 517 build backend

This patch creates a thin wrapper around the `setuptools`' PEP 517
build backend in-tree. It features an ability to request generating
the manpage files in the process of building a source distribution.
This toggle is implemented using the `config_settings` mechanism of
PEP 517.
One must explicitly pass it a CLI option to the build front-end to
trigger said behavior. The packagers are expected to use the
following call:

    python -m build --config-setting=--build-manpages

This option has no effect on building wheels.

🧪 The change includes integration tests

This test runs building and re-building sdists and wheels with and
without the `--build-manpages` config setting under the
oldest-supported and new `setuptools` pinned.

It is intended to preserve the interoperability of the packaging setup
across Python runtimes.

An extra smoke test also verifies that non PEP 517 interfaces remain functional.

PR #79606

Co-authored-by: Matt Clay <matt@mystile.com>
(cherry picked from commit 56036013cd09db0f1b20402a69be44345ddafb22)

webknjaz committed:
[backport-2.12] 🎨 Convert RST to manpage in-memory @ PEP 517

* Add _convert_rst_in_template_to_manpage arg types

* 📦 Make manpage build dependencies conditional

Previously, said dependencies were declared as unconditionally
required even when manpages not needed to be built. This patch
Makes it so they are only required when needed.

* Correct _generate_rst_in_templates returned type

It was marked as Path before this patch but in fact, it's iterable of
paths.

* 🎨 Convert RST to manpage in-memory @ PEP 517

Previously, the automation was writing a temporary templated RST on
disk and calling a helper CLI script on that. But with this change, it
happens with less unnecessary I/O.

Co-Authored-By: Matt Davis <6775756+nitzmahone@users.noreply.github.com>

* 📦Expose sdist manpage build deps unconditionally

Due to a bug in pypa/build, the `get_requires_for_build_sdist()` hook
is always invoked with `config_settings=None`. This means that we
cannot conditionally extend build requirements in said hook.

As a workaround, this patch makes hook pretend that `--built-manpages`
is always passed.

Ref: https://github.com/pypa/build/issues/559.

---------

Co-authored-by: Matt Davis <6775756+nitzmahone@users.noreply.github.com>
(cherry picked from commit 67bafafbc0d61751d2ec10b0aeccb6b01482a074)

This is a backport of #80098.

(cherry picked from commit 4ebf6c1c067a91a88c6c1dda69557cea703cd9ce)

mattclay commented on test/integration/targets/canonical-pep517-self-packaging/minimum-build-constraints.txt:
```diff
@@ -0,0 +1,16 @@
+# Lowest supporting Python 3.9 and 3.10:
+setuptools == 57.0.0; python_version == "3.9" or python_version == "3.10"
```
This will need to be updated to support Python 3.8.

mattclay reviewed (commented):
Aside from the Python 3.8 issue, there's a `package-data` sanity test failure. I suspect it has something to do with the fact we didn't backport https://github.com/ansible/ansible/pull/76021

webknjaz committed:
Drop `str.removeprefix()` use from PEP 517 backend

This is needed to support Python 3.8

(cherry picked from commit 287198bf5af7f3c04c3f2df36354edc39e37d9fe)

webknjaz committed:
Declare minimum setuptools @ PEP 517 backend test

(cherry picked from commit 501b2115cac3a8472e7d5e40232acbb95aad19de)

webknjaz committed:
[backport-2.12] Make isolated source changes @ PEP 517 backend

PR #80253

(cherry picked from commit 888abf5d6e13880ba82d02c5f8a57ca1f90a992f)
(cherry picked from commit 1a11b783499170a7dfec2e91ef13df422d3f8b59)

webknjaz committed:
[backport-2.12] 📦 Switch sdist build-system to pure setuptools

This patch modifies the in-tree build backend to build sdists that swap
out pointers to it in the `pyproject.toml`'s `[build-system]` section.

The effect of this is that the first build from source (for example,
from a Git checkout) uses our PEP 517 in-tree build backend. But the
produced tarball has `build-backend` set to `setuptools.build_meta`
which is the native build backend of `setuptools`. So any following
builds from that sdist will skip using the in-tree build backend,
calling the setuptools' one.
The good news is that if the first build generated the manpages, they
will be included and won't go anywhere even though, a different build
system is in place.

Combined with #80253, this will make sure not to modify the current
source checkout on that first build.

Co-authored-by: Matt Clay <matt@mystile.com>

PR #80255

(cherry picked from commit 7097df3eed979446830fc579613ffb9b7e7c19bf)
(cherry picked from commit ea5785c2fd7000a6bbe11628941d00110b3af019)

webknjaz committed:
[backport-2.12] Make PEP 517 mutation tests use pinned old setuptools

* Pin setuptools to lowest supported @ PEP 517 test

This allows catching the behavior of builds under old setuptools.

* Stop invoking `setup.py install` in tests

This is not the part we care about since it involves dealing with the
external runtime dependencies rather than building our source
distribution.

PR #80262

(cherry picked from commit eebfd71a6da396b011e664604c9991543205d780)
(cherry picked from commit ba41769f9a7a542583922703360cf6255d0d5812)

webknjaz committed:
Set straight.plugin in PEP 517 test to v1.4.2

This is needed since 1.5.0 is older, does not have a known source and
does not build under Python 3.8.

(cherry picked from commit b3dde9033a460c3fe311dd5d7ca0d529b373da86)

webknjaz committed:
Skip comparing `PKG-INFO` under Python 3.8 in test

This is necessary because setuptools is pinned to a very old version
in said environment and generates a substantially different output
file that is rather hard to normalize to a comparible state.

(cherry picked from commit 7369a1a4a0fc34d0ffdb59b59d18c1fa14f7ab8c)

webknjaz committed:
Make PEP 517 test compatible w/ py3.8 @ freebsd

This patch pre-compiles the wheel cache for MarkupSafe and straight.plugin
outside the temporary isolated PEP 517 build environment. It also pins
PyYAML to v5.4 below Python 3.9.

This is necessary to work around the build failures for the build
requirements that we don't particularly care about in said test.

Co-authored-by: Matt Clay <matt@mystile.com>
(cherry picked from commit d24e68fe5d293429fe6d96da411db22c7fa0224d)

webknjaz committed:
Upgrade pip to v20+ in the PEP 660 test

This is needed to ensure that pip in test supports in-tree build
backend.

(cherry picked from commit a4f5c99e4fac3cb37f6db873343427eb48ad36b3)

webknjaz committed:
📦 Include `pyproject.toml` into sdist explicitly

webknjaz committed:
📦 Hardcode version importable @ PEP 517 backend

mattclay commented:
Closing as 2.12 is now EOL upstream.
